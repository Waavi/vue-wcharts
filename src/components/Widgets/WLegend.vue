<template>
    <a
        class="WLegend"
        :title="text"
        :class="{ selected, active }"
        :style="styles"
        @click.prevent="$emit('onClick', { text, index })"
        @mouseenter="$emit('onMouseenter', { text, index })"
        @mouseleave="$emit('onMouseleave', { text, index })"
    >
        <slot
            name="bullet"
            :index="index"
            :text="text"
            :color="colorCmp"
        >
            <div class="Bullet">
                <span
                    class="Circle"
                    :style="{ backgroundColor: colorCmp }"
                />
            </div>
        </slot>
        <span class="Text">{{ text }}</span>
    </a>
</template>

<script>
import VueTypes from 'vue-types'
import themeMixin from '../../mixins/theme'

export default {
    name: 'WLegend',
    mixins: [themeMixin],
    props: {
        index: VueTypes.oneOfType([Number, String]),
        text: VueTypes.string,
        active: VueTypes.bool.def(false),
        selected: VueTypes.bool.def(false),
        styles: VueTypes.object,
        textStyles: VueTypes.object,
        color: VueTypes.string,
    },
    computed: {
        // Get color value. Color prop, Color theme by index or color by defualt
        colorCmp () {
            if (this.color) return this.color
            if (this.index) return this.colors[this.index]
            return this.colors[0]
        },
    },
}
</script>

<style scoped lang="scss">
.WLegend {
    display: flex;
    font-size: 12px;
    opacity: .5;
    transition: opacity .3s ease;
    cursor: pointer;

    &.selected, &.active {
        opacity: 1;
    }
}

.Bullet {
    display: inline-block;

    .Circle {
        display: inline-block;
        height: 10px;
        width: 10px;
        border-radius: 50%;
        margin-right: .5rem;
    }
}
</style>
