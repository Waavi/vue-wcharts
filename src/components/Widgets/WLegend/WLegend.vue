<template>
    <div
        v-if="!!Chart.legends.length"
        class="WLegend"
        :style="stylesCmp.component"
    >
        <ul
            class="Wrapper"
            :style="stylesCmp.wrapper"
        >
            <li
                v-for="(legend, index) in Chart.legends"
                :key="legend"
                class="Legend"
                :style="stylesCmp.legend"
            >
                <WLegendItem
                    :key="index"
                    :index="index"
                    :text="legend"
                    :active="!selectable ? false : Chart.activeElements.includes(index)"
                    :style="legendStylesCmp"
                    :disabledStyles="legendStylesDisabledCmp"
                    :colors="colors"
                    @onClick="handleClick"
                >
                    <template #bullet="{ index, text, color }">
                        <slot
                            name="bullet"
                            :index="index"
                            :text="text"
                            :color="color"
                        >
                            <WBullet :styles="{ ...bulletStylesCmp, borderColor: color }" />
                        </slot>
                    </template>
                </WLegendItem>
            </li>
        </ul>
    </div>
</template>

<script>
import VueTypes from 'vue-types'
import sortBy from 'lodash.sortby'
import { toPx, getIsHorizontal, getSpacesByPos } from '../utils'
import WLegendItem from '../WLegendItem/WLegendItem.vue'
import WBullet from '../WBullet/WBullet.vue'
import themeMixin from '../../../mixins/theme'

export default {
    name: 'WLegend',
    type: 'plugins',
    inject: ['Chart'],
    components: {
        WLegendItem,
        WBullet,
    },
    mixins: [themeMixin],
    props: {
        position: VueTypes.oneOf(['top', 'bottom', 'left', 'right']).def('bottom'),
        align: VueTypes.oneOf(['start', 'center', 'end']).def('center'),
        space: VueTypes.arrayOf(VueTypes.number).def([16, 16, 16, 16]),
        size: VueTypes.number, // Width or height, with different positiion prop top-bottom/left-right
        selectable: VueTypes.bool.def(false),
        colors: VueTypes.arrayOf(VueTypes.string),
        // Styles
        styles: VueTypes.object.def({}),
        wrapperStyles: VueTypes.object.def({}),
        legendStyles: VueTypes.object.def({}),
        legendStylesDisabled: VueTypes.object.def({}),
        bulletStyles: VueTypes.object.def({}),
    },
    preload ({ parent, props, index }) {
        const { position, size, space } = props
        // Positions
        const isHorizontal = getIsHorizontal(position)
        const innerWidth = !isHorizontal ? size || 85 : null
        const innerHeight = isHorizontal ? size || 20 : null
        // Spaces
        const [top, right, bottom, left] = space || this.props.space.default()
        const width = right + left + innerWidth
        const height = top + bottom + innerHeight
        // Resize parent
        const spaces = getSpacesByPos(position, { width, height }, parent.spaceObjects)
        parent.addSpace(spaces)
    },
    computed: {
        stylesCmp () {
            const { position, align, size } = this
            const isHorizontal = getIsHorizontal(position)
            // Positions
            const width = isHorizontal ? '100%' : size || toPx(85)
            const height = !isHorizontal ? '100%' : size || toPx(20)
            // Spaces
            const [topParent, rightParent, bottomParent, leftParent] = this.Chart.space
            const [top, right, bottom, left] = this.space
            const spaces = {
                top, bottom, left, right,
            }
            // Styles
            const justifyContent = align !== 'center' ? `flex-${align}` : align
            const aligns = {
                vertical: { flexDirection: 'column', padding: [toPx(topParent), 0, toPx(bottomParent), 0].join(' ') },
                horizontal: { flexDirection: 'row', padding: [0, toPx(rightParent), 0, toPx(leftParent)].join(' ') },
            }

            return {
                component: {
                    top: !isHorizontal ? 0 : null,
                    [position]: toPx(spaces[position]),
                    width,
                    height,
                    ...this.themeStyles.styles,
                    ...this.styles,
                },
                wrapper: {
                    ...aligns[isHorizontal ? 'horizontal' : 'vertical'],
                    justifyContent,
                    ...this.themeStyles.wrapperStyles,
                    ...this.wrapperStyles,
                },
                legend: {
                    margin: !isHorizontal ? '.5rem 0 0 0' : '0 0 0 .75rem',
                },
            }
        },
        legendStylesCmp () {
            return {
                ...this.themeStyles.legendStyles,
                ...this.legendStyles,
            }
        },
        legendStylesDisabledCmp () {
            return {
                ...this.themeStyles.legendStylesDisabled,
                ...this.legendStylesDisabled,
            }
        },
        bulletStylesCmp () {
            return {
                ...this.themeStyles.bulletStyles,
                ...this.bulletStyles,
            }
        },
    },
    methods: {
        /*
        *  Handle on click legend, if has selectable prop true
        */
        handleClick ({ legend, index }) {
            if (!this.selectable) return
            let actives = [...this.Chart.activeElements]

            if (!actives.includes(index)) actives.push(index)
            else actives = actives.filter(n => n !== index)

            this.Chart.activeElements = sortBy(actives)
        },
    },
}
</script>
