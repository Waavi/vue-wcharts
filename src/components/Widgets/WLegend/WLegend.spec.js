import { shallowMount } from '@vue/test-utils'
import WLegend from './WLegend.vue'

describe('Widgets/WLegend', () => {
    const customStyles = {
        backgroundColor: '#000',
    }
    const propsData = {
        position: 'top',
        align: 'start',
        selectable: true,
        space: [20, 20, 20, 20],
        styles: customStyles,
        wrapperStyles: customStyles,
        legendStyles: customStyles,
        bulletStyles: customStyles,
    }

    const provide = {
        Chart: {
            legends: ['One Bar', 'Two Bar'],
            space: [20, 20, 20, 20],
            activeElements: [0, 1],
        },
    }

    const provideWithoutLegends = {
        Chart: {
            legends: [],
            space: [20, 20, 20, 20],
            activeElements: [],
        },
    }

    const defaultConfig = {
        propsData,
        provide,
    }

    it(`Should be render correctly`, () => {
        const wrapper = shallowMount(WLegend, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Shouldn't be render`, () => {
        const wrapper = shallowMount(WLegend, { ...defaultConfig, provide: provideWithoutLegends })
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be handle activeElements in Chart after click WLegendItem`, () => {
        const wrapper = shallowMount(WLegend, defaultConfig)
        wrapper.vm.handleClick({ legend: undefined, index: 0 })
        expect(wrapper.vm.Chart.activeElements).toEqual([1])
    })

    it(`Should be render correctly, without space of Chart`, () => {
        const wrapper = shallowMount(WLegend, {
            ...defaultConfig,
            provide: {
                Chart: {
                    ...provide.Chart,
                    space: undefined,
                },
            },
        })
        expect(wrapper.vm.chartSpace).toEqual([0, 0, 0, 0])
    })

    it(`Should be render correctly, with custom styles and position horizontal`, () => {
        const wrapper = shallowMount(WLegend, {
            ...defaultConfig,
            propsData: {
                ...propsData,
                position: 'left',
            },
        })
        expect(wrapper.vm.stylesCmp.component.width).toEqual('85px')
    })
})
