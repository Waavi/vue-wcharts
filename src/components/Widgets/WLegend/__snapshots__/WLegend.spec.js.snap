// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widgets/WLegend Should be render correctly 1`] = `
<div class="WLegend" style="top: 20px; width: 100%; height: 20px; position: absolute; display: flex; background-color: rgb(0, 0, 0);">
  <ul class="Wrapper" style="flex-direction: row; padding: 0px 20px 0px 20px; justify-content: flex-start; display: flex; flex: 1; margin: 0px; list-style: none; background-color: rgb(0, 0, 0);">
    <li class="Legend" style="margin: 0px 0px 0px .75rem;">
      <wlegenditem-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" animated="true" transduration="1" transeffect="ease" index="0" text="One Bar" active="true" color="" styles="[object Object]" noactivestyles="[object Object]" bulletstyles="[object Object]" style="display: flex; align-items: center; font-size: 12px; cursor: pointer; background-color: rgb(0, 0, 0);"></wlegenditem-stub>
    </li>
    <li class="Legend" style="margin: 0px 0px 0px .75rem;">
      <wlegenditem-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" animated="true" transduration="1" transeffect="ease" index="1" text="Two Bar" active="true" color="" styles="[object Object]" noactivestyles="[object Object]" bulletstyles="[object Object]" style="display: flex; align-items: center; font-size: 12px; cursor: pointer; background-color: rgb(0, 0, 0);"></wlegenditem-stub>
    </li>
  </ul>
</div>
`;

exports[`Widgets/WLegend Shouldn't be render 1`] = `undefined`;
